File: framework/lib/textbox.dart
part of uxml;

/**
 * Implements textbox control to display editable text.
 *
 * @author ferhat@ (Ferhat Buyukkokten)
 */
class TextBox extends Control {

  static ElementDef textboxElementDef;
  /** Text property definition */
  static PropertyDefinition textProperty;
  /** Wordwrap property definition */
  static PropertyDefinition wordWrapProperty;
  /** HtmlEnabled property definition */
  static PropertyDefinition htmlEnabledProperty;
  /** Multiline property definition */
  static PropertyDefinition multilineProperty;

  /** Display as password property definition */
  static PropertyDefinition displayAsPasswordProperty;
  /** Readonly property definition */
  static PropertyDefinition readOnlyProperty;
  /** IsEmpty property definition */
  static PropertyDefinition isEmptyProperty;
  /** MaxChars property definition */
  static PropertyDefinition maxCharsProperty;
  /** CharsRemaining property definition */
  static PropertyDefinition charsRemainingProperty;
  /** TextChanged event definition */
  static EventDef textChangedEvent;

  static EventArgs _changeEventArgs;
  bool _deferredFocus;

  TextBox() : super() {
    focusEnabled = true;
    _deferredFocus = false;
  }

  /**
   * Sets or returns text.
   */
  String get text => getProperty(textProperty);

  set text(String value) {
    setProperty(textProperty, value);
  }

  /**
   * Sets or returns displayAsPassword.
   */
  bool get displayAsPassword => getProperty(displayAsPasswordProperty);

  set displayAsPassword(bool value) {
    setProperty(displayAsPasswordProperty, value);
  }

  /**
   * Sets or returns if multiline is enabled.
   */
  bool get multiline => getProperty(multilineProperty);

  set multiline(bool value) {
    setProperty(multilineProperty, value);
  }

  /**
   * Sets or returns if wordWrap is enabled.
   */
  bool get wordWrap => getProperty(wordWrapProperty);

  set wordWrap(bool value) {
    setProperty(wordWrapProperty, value);
  }

  /**
   * Sets or returns if text content is html.
   */
  bool get htmlEnabled => getProperty(htmlEnabledProperty);

  set htmlEnabled(bool value) {
    setProperty(htmlEnabledProperty, value);
  }

  /**
   * Sets or returns if text content is editable.
   */
  bool get readonly => getProperty(readOnlyProperty);

  set readonly(bool value) {
    setProperty(readOnlyProperty, value);
  }

  /**
   * Sets or returns max chars allowed in text field.
   */
  int get maxChars => getProperty(maxCharsProperty);

  set maxChars(int value) {
    setProperty(maxCharsProperty, value);
  }

  /** Overrides UIElement.onChromeChanged. */
  void onChromeChanged(Chrome oldChrome, Chrome newChrome) {
    super.onChromeChanged(oldChrome, newChrome);
    _redirectFocusChrome();
  }

  /** @see UIElement.initSurface. */
  void initSurface(UISurface parentSurface, [int index = -1]) {
    super.initSurface(parentSurface, index);
    _redirectFocusChrome();
    if (_deferredFocus && (_internalTextEdit != null)) {
      _internalTextEdit.setFocus();
    }
  }

  TextEdit get _internalTextEdit {
    if (chromeTree != null) {
      return chromeTree.getElementByType(TextEdit.texteditElementDef);
    }
    return null;
  }

  void _redirectFocusChrome() {
    if (chromeTree != null) {
      TextEdit textEdit = _internalTextEdit;
      if (textEdit != null) {
        FocusManager.redirectFocusChrome(textEdit, this);
      }
    }
  }

  /** Overrides UIElement.onMouseDown. */
  void onMouseDown(MouseEventArgs mouseArgs) {
    TextEdit textEdit = _internalTextEdit;
    if (textEdit != null) {
      if (textEdit.isFocused == false) {
        textEdit.setFocus();
      }
    }
    mouseArgs.handled = true;
    mouseArgs._passthrough();
  }

  /**
   * Selects all text in editor.
   */
 void selectAll() {
    if (_internalTextEdit != null) {
      _internalTextEdit.selectAll();
    }
  }

  /**
   * Returns the number of chars remaining out of maxChars.
   */
  int get charsRemaining => getProperty(charsRemainingProperty);

  /**
   * Returns true if textbox is empty.
   */
  bool get isEmpty => getProperty(isEmptyProperty);

  static void _maxCharsChangedHandler(Object target,
                                      Object propDef,
                                      Object oldValue,
                                      Object newValue) {
    TextBox textBox = target;
    textBox._updateCharsRemaining();
  }

  static void _textChangedHandler(Object target,
                                  Object propDef,
                                  Object oldValue,
                                  Object newValue) {
    TextBox textBox = target;
    if (newValue != null && textBox.maxChars > 0) {
      textBox._updateCharsRemaining();
    }

    _changeEventArgs.source = textBox;
    textBox.notifyListeners(textChangedEvent, _changeEventArgs);
  }

  void _updateCharsRemaining() {
    setProperty(isEmptyProperty, text.length == 0);
    if (maxChars > 0) {
      setProperty(charsRemainingProperty, maxChars - text.length);
    } else {
      setProperty(charsRemainingProperty, 0x7FFFFFFF);
    }
  }

  /** Sets input focus to textbox. */
  void setFocus() {
    if (_hostSurface == null) {
      _deferredFocus = true;
    }
    TextEdit textEdit = _internalTextEdit;
    if (textEdit != null) {
      textEdit.setFocus();
    }
  }

  /** @see UxmlElement.getDefinition. */
  ElementDef getDefinition() => textboxElementDef;

  /** Registers component. */
  static void registerTextBox() {
    textProperty = ElementRegistry.registerProperty("text",
        PropertyType.STRING, PropertyFlags.RESIZE, _textChangedHandler, "");
    wordWrapProperty = ElementRegistry.registerProperty("wordWrap",
        PropertyType.BOOL, PropertyFlags.RESIZE, null, false);
    htmlEnabledProperty = ElementRegistry.registerProperty("htmlEnabled",
        PropertyType.BOOL, PropertyFlags.RESIZE, null, false);
    multilineProperty = ElementRegistry.registerProperty("multiline",
        PropertyType.BOOL, PropertyFlags.RESIZE, null, false);
    displayAsPasswordProperty = ElementRegistry.registerProperty(
        "displayAsPassword", PropertyType.BOOL, PropertyFlags.NONE, null,
        false);
    readOnlyProperty = ElementRegistry.registerProperty("ReadOnly",
        PropertyType.BOOL, PropertyFlags.NONE, null, false);
    isEmptyProperty = ElementRegistry.registerProperty("IsEmpty",
        PropertyType.BOOL, PropertyFlags.NONE, null, true);
    maxCharsProperty = ElementRegistry.registerProperty("MaxChars",
        PropertyType.INT, PropertyFlags.NONE, _maxCharsChangedHandler, 0);
    charsRemainingProperty = ElementRegistry.registerProperty("CharsRemaining",
        PropertyType.INT, PropertyFlags.NONE, null, 0x7FFFFFFF);
    textChangedEvent = new EventDef("TextChanged", Route.DIRECT);
    textboxElementDef = ElementRegistry.register("TextBox",
        Control.controlElementDef,
        [textProperty, wordWrapProperty, htmlEnabledProperty, isEmptyProperty,
        multilineProperty, displayAsPasswordProperty, readOnlyProperty,
        maxCharsProperty, charsRemainingProperty], null);
    Cursor.cursorProperty.overrideDefaultValue(textboxElementDef, Cursor.IBEAM);
    _changeEventArgs = new EventArgs(null);
  }
}

File: plugin/com.hello.uxml.tools.ui/src/com/hello/uxml/tools/framework/TextBox.java
package com.hello.uxml.tools.framework;

import com.hello.uxml.tools.framework.events.EventArgs;
import com.hello.uxml.tools.framework.events.EventDefinition;
import com.hello.uxml.tools.framework.events.EventManager;

import java.util.EnumSet;

/**
 * Implements a stylable TextEdit control.
 *
 * @author ferhat@(Ferhat Buyukkokten)
 */
public class TextBox extends Control {

  /** Text Property Definition */
  public static PropertyDefinition textPropDef = PropertySystem.register("Text",
      String.class,
      TextBox.class,
      new PropertyData("", EnumSet.of(PropertyFlags.Resize, PropertyFlags.Redraw,
          PropertyFlags.Localizable),
          new PropertyChangeListener() {
        @Override
        public void propertyChanged(PropertyChangedEvent e) {
          TextBox element = (TextBox) e.getSource();
          EventArgs eArgs = new EventArgs(element, textChangedEvent);
          element.raiseEvent(eArgs);
          }
        }
      ));

  /**
   * DisplayAsPassword property definition.
   */
  public static PropertyDefinition displayAsPasswordPropDef = PropertySystem.register(
      "DisplayAsPassword", Boolean.class, TextBox.class,
      new PropertyData(true, EnumSet.of(PropertyFlags.None)));

  /**
   * Multiline property definition.
   */
  public static PropertyDefinition multilinePropDef = PropertySystem.register(
      "Multiline", Boolean.class, TextBox.class,
      new PropertyData(false, EnumSet.of(PropertyFlags.None)));

  /**
   * WordWrap property definition.
   */
  public static PropertyDefinition wordWrapPropDef = PropertySystem.register(
      "WordWrap", Boolean.class, TextBox.class,
      new PropertyData(true, EnumSet.of(PropertyFlags.None)));

  /**
   * HtmlEnabled property definition.
   */
  public static PropertyDefinition htmlEnabledPropDef = PropertySystem.register(
      "HtmlEnabled", Boolean.class, TextBox.class,
      new PropertyData(true, EnumSet.of(PropertyFlags.None)));

  /**
   * Readonly property definition.
   */
  public static PropertyDefinition readOnlyPropDef = PropertySystem.register(
      "ReadOnly", Boolean.class, TextBox.class,
      new PropertyData(true, EnumSet.of(PropertyFlags.None), null));

  /**
   * MaxChars property definition.
   */
  public static PropertyDefinition maxCharsPropDef = PropertySystem.register(
      "MaxChars", Integer.class, TextBox.class,
      new PropertyData(0, EnumSet.of(PropertyFlags.None), null));

  /** TextChanged event definition */
  public static EventDefinition textChangedEvent = EventManager.register(
      "TextChanged", TextBox.class, EventArgs.class, null, null);

  /**
   * Returns text.
   */
  public String getText() {
    return (String) getProperty(textPropDef);
  }

  /**
   * Sets text.
   */
  public void setText(String text) {
    setProperty(textPropDef, text);
  }

  /**
  * Sets/returns displayAsPassword property.
  */
  public void setDisplayAsPassword(boolean value) {
    setProperty(displayAsPasswordPropDef, value);
  }

  public boolean getDisplayAsPassword() {
    return ((Boolean) getProperty(displayAsPasswordPropDef));
  }

  /**
   * Sets/returns if multiline is enabled.
   */
   public void setMultiline(boolean value) {
     setProperty(multilinePropDef, value);
   }

   public boolean getMultiline() {
     return ((Boolean) getProperty(multilinePropDef));
   }

   /**
    * Sets/returns if word wrapping is enabled.
    */
   public void setWordWrap(boolean value) {
     setProperty(wordWrapPropDef, value);
   }

   public boolean getWordWrap() {
     return ((Boolean) getProperty(wordWrapPropDef));
   }

   /**
    * Sets/returns if editor is readonly.
    */
   public void setReadOnly(boolean value) {
     setProperty(readOnlyPropDef, value);
   }

   public boolean getReadOnly() {
     return ((Boolean) getProperty(readOnlyPropDef));
   }

   /**
    * Sets/returns max chars allowed in the textedit.
    */
   public void setMaxChars(int value) {
     setProperty(maxCharsPropDef, value);
   }

   public int getMaxChars() {
     return ((Integer) getProperty(maxCharsPropDef));
   }

   /**
    * Sets/returns if text content is html.
    */
   public void setHtmlEnabled(boolean value) {
     setProperty(htmlEnabledPropDef, value);
   }

   public boolean getHtmlEnabled() {
     return ((Boolean) getProperty(htmlEnabledPropDef));
   }
}
